File: ch11/MyNotes/PersistenceLayer/dao/NoteDAO.m
//
//  NoteDAO.m
//  MyNotes
//
//  Created by MyZiyue on 2018/12/29.
//  Copyright © 2018 MyZiyue.com. All rights reserved.
//

#import "NoteDAO.h"

@implementation NoteDAO

static NoteDAO *sharedSingleton = nil;

+(NoteDAO*)sharedInstance {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedSingleton = [[self alloc] init];
        
        NSDateFormatter* dateFormatter = [[NSDateFormatter alloc] init];
        [dateFormatter setDateFormat:@"yyyy-MM-dd HH:mm:ss"];
        
        NSDate* date1 = [dateFormatter dateFromString:@"2016-01-01 16:01:03"];
        Note* note1 = [[Note alloc] initWithDate:date1 content:@"Welcome to MyNote."];
        
        NSDate* date2 = [dateFormatter dateFromString:@"2016-01-02 08:01:03"];
        Note* note2 = [[Note alloc] initWithDate:date2 content:@"Welcome to MyNote."];
        
        sharedSingleton.listData = [[NSMutableArray alloc] init];
        [sharedSingleton.listData addObject:note1];
        [sharedSingleton.listData addObject:note2];
    });
    return sharedSingleton;
}

-(int)create:(Note*)model {
    [self.listData addObject:model];
    return 0;
}

-(int)remove:(Note*)model {
    for (Note* note in self.listData) {
        if([note.date isEqualToDate:model.date]) {
            [self.listData removeObject:note];
            break;
        }
    }
    return 0;
}

-(int) modify:(Note*)model {
    for (Note* note in self.listData) {
        //比较日期主键是否相等
        if ([note.date isEqualToDate:model.date]){
            note.content = model.content;
            break;
        }
    }
    return 0;
}

-(NSMutableArray*) findAll {
    return self.listData;
}

-(Note*) findById:(Note*)model {
    for (Note* note in self.listData) {
        //比较日期主键是否相等
        if ([note.date isEqualToDate:model.date]){
            return note;
        }
    }
    return nil;
}


@end

File: ch11/SMyNotes/PersistenceLayer/dao/NoteDAO.swift
//
//  NoteDAO.swift
//  SMyNotes
//
//  Created by MyZiyue on 2018/12/29.
//  Copyright © 2018 MyZiyue.com. All rights reserved.
//

import Foundation

class NoteDAO {
    var listData: NSMutableArray!
    static let sharedInstance: NoteDAO = {
        let instance = NoteDAO()
        
        let dateFormatter: DateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-DD HH:mm:ss"
        let date1: NSDate = dateFormatter.date(from: "2016-01-01 16:01:03") as! NSDate
        let note1: Note = Note(date: date1, content: "Welcome to MyNote.")
        
        let date2: NSDate = dateFormatter.date(from: "2016-01-02 08:01:03") as! NSDate
        let note2: Note = Note(date: date2, content: "欢迎使用MyNote。")
        
        instance.listData = NSMutableArray()
        instance.listData.addObjects(note1)
        instance.listData.addObjects(note2)
        
        return instance
    }()
    
    func create(model: Note) -> Int {
        self.listData.addObjects(model)
        return 0
    }
    
    func remove(model: Note) -> Int {
        for note in self.listData {
            var note2 = note as! Note
            // 比较日期主键是否相等
            if note2.date == model.date {
                self.listData.removeObject(note2)
                break
            }
        }
        return 0
    }
    
    func modify(model: Note) -> Int {
        for note in self.listData {
            var note2 = note as! Note
            if note2.date == model.date {
                note2.content = model.content
                break
            }
        }
        return 0
    }
    
    func findAll() -> NSMutableArray {
        return self.listData
    }
    
    func findById(model: Note) -> Note? {
        for note in self.listData {
            var note2 = note as Note
            if note2.date == model.date {
                return note2
            }
        }
        return nil
    }
}
